certora/specs/Reverts.spec

```diff
@@ -0,0 +1,179 @@
+
+ghost mapping(MorphoHarness.Id => mathint) sumCollateral
+{
+    init_state axiom (forall MorphoHarness.Id id. sumCollateral[id] == 0);
+}
+hook Sstore position[KEY MorphoHarness.Id id][KEY address owner].collateral uint128 newAmount (uint128 oldAmount) STORAGE {
+    sumCollateral[id] = sumCollateral[id] - oldAmount + newAmount;
+}
+
+definition emptyMarket(MorphoHarness.Id id) returns bool =
+    totalSupplyAssets(id) == 0 &&
+    totalSupplyShares(id) == 0 &&
+    totalBorrowAssets(id) == 0 &&
+    totalBorrowShares(id) == 0 &&
+    sumCollateral[id] == 0;
+
+definition exactlyOneZero(uint256 assets, uint256 shares) returns bool =
+    (assets == 0 && shares != 0) || (assets != 0 && shares == 0);
+
+// This invariant catches bugs when not checking that the market is created with lastUpdate.
+invariant notInitializedEmpty(MorphoHarness.Id id)
+    !isCreated(id) => emptyMarket(id)
+{
+    preserved with (env e) {
+        // Safe require because timestamps cannot realistically be that large.
+        require e.block.timestamp < 2^128;
+    }
+}
+
+// Useful to ensure that authorized parties are not the zero address and so we can omit the sanity check in this case.
+invariant zeroDoesNotAuthorize(address authorized)
+    !isAuthorized(0, authorized)
+{
+    preserved setAuthorization(address _authorized, bool _newAuthorization) with (env e) {
+        // Safe require because no one controls the zero address.
+        require e.msg.sender != 0;
+    }
+}
+
+// Check the revert condition for the setOwner function.
+rule setOwnerRevertCondition(env e, address newOwner) {
+    address oldOwner = owner();
+    setOwner@withrevert(e, newOwner);
+    assert lastReverted <=> e.msg.value != 0 || e.msg.sender != oldOwner || newOwner == oldOwner;
+}
+
+// Check the revert condition for the setOwner function.
+rule enableIrmRevertCondition(env e, address irm) {
+    address oldOwner = owner();
+    bool oldIsIrmEnabled = isIrmEnabled(irm);
+    enableIrm@withrevert(e, irm);
+    assert lastReverted <=> e.msg.value != 0 || e.msg.sender != oldOwner || oldIsIrmEnabled;
+}
+
+// Check the revert condition for the enableLltv function.
+rule enableLltvRevertCondition(env e, uint256 lltv) {
+    address oldOwner = owner();
+    bool oldIsLltvEnabled = isLltvEnabled(lltv);
+    enableLltv@withrevert(e, lltv);
+    assert lastReverted <=> e.msg.value != 0 || e.msg.sender != oldOwner || lltv >= wad() || oldIsLltvEnabled;
+}
+
+// Check that setFee reverts when its inputs are not validated.
+// setFee can also revert if the accrueInterest reverts.
+rule setFeeInputValidation(env e, MorphoHarness.MarketParams marketParams, uint256 newFee) {
+    MorphoHarness.Id id = libId(marketParams);
+    address oldOwner = owner();
+    bool wasCreated = isCreated(id);
+    setFee@withrevert(e, marketParams, newFee);
+    bool hasReverted = lastReverted;
+    assert e.msg.value != 0 || e.msg.sender != oldOwner || !wasCreated || newFee > maxFee() => hasReverted;
```

Comment:
It does not work because `lastReverted` will be the revert status of the last call, so `maxFee()` in this case which will never revert. I suggested to the Certora team that only calls with the suffix `@withrevert` should be considered for the variable `lastReverted`, but this is a breaking change so they have not yet considered it